core/src/main/java/nu/yona/server/crypto/CryptoSession.java

```diff
@@ -136,36 +129,34 @@ public static CryptoSession getCurrent()
 		return cryptoSession;
 	}
 
+	/**
+	 * Encrypts the given plaintext bytes.
+	 * 
+	 * @param plaintext the bytes to be encrypted
+	 * @return the encrypted bytes, with the crypto variant number prepended to it.
+	 */
 	public byte[] encrypt(byte[] plaintext)
 	{
-		try
-		{
-			return getEncryptionCipher().doFinal(plaintext);
-		}
-		catch (IllegalBlockSizeException | BadPaddingException e)
-		{
-			throw CryptoException.encryptingData(e);
-		}
+		return CryptoUtil.encrypt(CURRENT_CRYPTO_VARIANT_NUMBER, getEncryptionCipher(), plaintext);
 	}
 
+	/**
+	 * Decrypts the given ciphertext bytes.
+	 * 
+	 * @param ciphertext the bytes to be decrypted, with a leading crypto variant number
+	 * @return the decrypted bytes
+	 */
 	public byte[] decrypt(byte[] ciphertext)
 	{
-		try
-		{
-			return getDecryptionCipher().doFinal(ciphertext);
-		}
-		catch (IllegalBlockSizeException | BadPaddingException e)
-		{
-			throw CryptoException.decryptingData(e);
-		}
+		return CryptoUtil.decrypt(CURRENT_CRYPTO_VARIANT_NUMBER, getDecryptionCipher(), ciphertext);
 	}
 
-	private SecretKey getSecretKey(char[] password)
+	private static SecretKey getSecretKey(String password)
 	{
 		try
 		{
-			SecretKeyFactory factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");
-			KeySpec spec = new PBEKeySpec(password, SALT, 65536, 128);
+			SecretKeyFactory factory = SecretKeyFactory.getInstance(SECRET_KEY_ALGORITHM);
+			KeySpec spec = new PBEKeySpec(password.toCharArray(), SALT, 65536, SECRET_KEY_LENGTH_BITS);
```

Comment:
This I'm not sure of. [Wikipedia](https://en.wikipedia.org/wiki/Key_derivation_function) talks about  "often tens or hundreds of thousands". Given our random passwords, it might be sufficient to go down. However, given that the performance tests till now indicate that we're fast enough, I'd rather not go down (this number is there since the first version).

I still intend to a security expert to review our solution. We have one in our network that already offered to look into it.